File: cli/internal/connectapi/attachments.go
package connectapi

import (
	"context"
	"strings"

	"connectrpc.com/connect"
	"hmans.de/chatto/internal/core"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
	corev1 "hmans.de/chatto/internal/pb/chatto/core/v1"
)

type attachmentService struct {
	api *API
}

type attachmentThumbnailRequest struct {
	width  int
	height int
	fit    string
}

func (s *attachmentService) ListRoomAttachments(ctx context.Context, req *connect.Request[apiv1.ListRoomAttachmentsRequest]) (*connect.Response[apiv1.ListRoomAttachmentsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	result, err := s.api.core.ListRoomAttachments(ctx, core.ListRoomAttachmentsInput{
		ActorID: caller.UserID,
		RoomID:  req.Msg.RoomId,
		Limit:   int(req.Msg.Limit),
		Offset:  int(req.Msg.Offset),
	})
	if err != nil {
		return nil, connectError(err)
	}

	thumbnail := attachmentThumbnailOptions(req.Msg.Thumbnail)
	items := make([]*apiv1.RoomAttachmentListItem, 0, len(result.Items))
	for _, item := range result.Items {
		if item == nil {
			continue
		}
		items = append(items, &apiv1.RoomAttachmentListItem{
			Attachment:        s.attachment(item.Attachment, caller.UserID, thumbnail),
			MessageEventId:    item.MessageEventID,
			ThreadRootEventId: item.ThreadRootEventID,
			CreatedAt:         item.CreatedAt,
		})
	}

	return connect.NewResponse(&apiv1.ListRoomAttachmentsResponse{
		Items:      items,
		TotalCount: int32(result.TotalCount),
		HasMore:    result.HasMore,
	}), nil
}

func (s *attachmentService) RefreshMessageAttachmentUrls(ctx context.Context, req *connect.Request[apiv1.RefreshMessageAttachmentUrlsRequest]) (*connect.Response[apiv1.RefreshMessageAttachmentUrlsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	attachments, err := s.api.core.MessageAttachments(ctx, core.MessageAttachmentsInput{
		ActorID: caller.UserID,
		RoomID:  req.Msg.RoomId,
		EventID: req.Msg.EventId,
	})
	if err != nil {
		return nil, connectError(err)
	}

	thumbnail := attachmentThumbnailOptions(req.Msg.Thumbnail)
	items := make([]*apiv1.RefreshedAttachmentUrls, 0, len(attachments))
	for _, attachment := range attachments {
		if attachment == nil {
			continue
		}
		video := (&timelineHydrator{
			api:      s.api,
			viewerID: caller.UserID,
		}).videoProcessing(attachment)
		items = append(items, &apiv1.RefreshedAttachmentUrls{
			AttachmentId:           attachment.Id,
			AssetUrl:               assetURLView(s.api.core.GetStableAttachmentAssetURL(attachment.Id, caller.UserID)),
			ThumbnailAssetUrl:      assetURLView(s.api.core.GetStableTransformedAttachmentAssetURL(attachment.Id, caller.UserID, thumbnail.width, thumbnail.height, thumbnail.fit)),
			VideoThumbnailAssetUrl: s.videoThumbnailAssetURL(attachment, caller.UserID),
			Variants:               refreshedVideoVariants(video),
		})
	}

	return connect.NewResponse(&apiv1.RefreshMessageAttachmentUrlsResponse{
		Attachments: items,
	}), nil
}

func refreshedVideoVariants(processing *apiv1.RoomTimelineVideoProcessing) []*apiv1.RoomTimelineVideoVariant {
	if processing == nil {
		return nil
	}
	return processing.GetVariants()
}

func (s *attachmentService) attachment(attachment *corev1.Attachment, viewerID string, thumbnail attachmentThumbnailRequest) *apiv1.RoomTimelineAttachment {
	if attachment == nil {
		return nil
	}
	h := &timelineHydrator{
		api:      s.api,
		viewerID: viewerID,
	}
	return &apiv1.RoomTimelineAttachment{
		Id:                attachment.Id,
		Filename:          attachment.Filename,
		ContentType:       attachment.ContentType,
		Width:             attachment.Width,
		Height:            attachment.Height,
		AssetUrl:          assetURLView(s.api.core.GetStableAttachmentAssetURL(attachment.Id, viewerID)),
		ThumbnailAssetUrl: assetURLView(s.api.core.GetStableTransformedAttachmentAssetURL(attachment.Id, viewerID, thumbnail.width, thumbnail.height, thumbnail.fit)),
		VideoProcessing:   h.videoProcessing(attachment),
	}
}

func (s *attachmentService) videoThumbnailAssetURL(attachment *corev1.Attachment, viewerID string) *apiv1.RoomTimelineAssetUrl {
	if attachment == nil || (!strings.HasPrefix(attachment.GetContentType(), "video/") && attachment.GetContentType() != "image/gif") {
		return nil
	}
	manifest, ok := s.api.core.Assets.VideoAttachmentManifest(attachment.GetId())
	if !ok || manifest == nil || manifest.Succeeded == nil || manifest.Succeeded.GetVideo() == nil {
		return nil
	}
	thumbnailID := manifest.Succeeded.GetVideo().GetThumbnailAssetId()
	if thumbnailID == "" {
		return nil
	}
	return assetURLView(s.api.core.GetStableAttachmentAssetURL(thumbnailID, viewerID))
}

func attachmentThumbnailOptions(options *apiv1.AttachmentThumbnailOptions) attachmentThumbnailRequest {
	width, height := 120, 120
	fit := "cover"
	if options != nil {
		if options.GetWidth() > 0 {
			width = int(options.GetWidth())
		}
		if options.GetHeight() > 0 {
			height = int(options.GetHeight())
		}
		switch options.GetFit() {
		case apiv1.AttachmentFitMode_ATTACHMENT_FIT_MODE_CONTAIN:
			fit = "contain"
		case apiv1.AttachmentFitMode_ATTACHMENT_FIT_MODE_COVER:
			fit = "cover"
		}
	}
	return attachmentThumbnailRequest{width: width, height: height, fit: fit}
}

File: apps/frontend/src/lib/api/attachments.ts
import { Code, ConnectError, createClient } from '@connectrpc/connect';
import { createConnectTransport } from '@connectrpc/connect-web';
import { FitMode } from '$lib/render/types';
import type { ExpiringAssetUrl, RefreshedAttachmentUrls } from '$lib/attachments/attachmentUrls';
import { serverRegistry } from '$lib/state/server/registry.svelte';
import { AttachmentService } from '@chatto/api-types/api/v1/attachments_connect';
import {
  AttachmentFitMode,
  AttachmentThumbnailOptions
} from '@chatto/api-types/api/v1/attachments_pb';
import {
  RoomTimelineVideoProcessingStatus,
  type RoomTimelineAssetUrl,
  type RoomTimelineAttachment,
  type RoomTimelineVideoProcessing
} from '@chatto/api-types/api/v1/room_timeline_pb';

export type AttachmentAPIConfig = {
  serverId?: string;
  baseUrl: string;
  bearerToken: string | null;
};

export type AttachmentRefreshOptions = {
  width: number;
  height: number;
  fit: FitMode;
};

export type RoomFileItem = {
  messageEventId: string;
  threadRootEventId: string | null;
  createdAt: string;
  attachment: {
    id: string;
    filename: string;
    contentType: string;
    width: number;
    height: number;
    assetUrl: ExpiringAssetUrl;
    thumbnailAssetUrl: ExpiringAssetUrl | null;
    videoProcessing: {
      status: 'PROCESSING' | 'COMPLETED' | 'FAILED';
      durationMs: number | null;
      width: number | null;
      height: number | null;
      sourceAvailable: boolean;
      reasonCode: string | null;
      thumbnailAssetUrl: ExpiringAssetUrl | null;
      variants: Array<{
        quality: string;
        width: number;
        height: number;
        size: number;
        assetUrl: ExpiringAssetUrl;
      }>;
    } | null;
  };
};

export type RoomFilesPage = {
  items: RoomFileItem[];
  totalCount: number;
  hasMore: boolean;
};

export type AttachmentAPI = {
  listRoomAttachments(input: {
    roomId: string;
    limit: number;
    offset: number;
    thumbnail: AttachmentRefreshOptions;
  }): Promise<RoomFilesPage>;
  refreshMessageAttachmentUrls(
    roomId: string,
    eventId: string,
    thumbnail: AttachmentRefreshOptions
  ): Promise<Map<string, RefreshedAttachmentUrls>>;
};

export function createAttachmentAPI(config: AttachmentAPIConfig): AttachmentAPI {
  const transport = createConnectTransport({
    baseUrl: config.baseUrl,
    useBinaryFormat: true
  });
  const client = createClient(AttachmentService, transport);
  const headers = () =>
    config.bearerToken ? { Authorization: `Bearer ${config.bearerToken}` } : undefined;

  async function handleAuthError(err: unknown): Promise<never> {
    if (err instanceof ConnectError && err.code === Code.Unauthenticated && config.serverId) {
      serverRegistry.handleAuthenticationRequired(config.serverId);
    }
    throw err;
  }

  return {
    async listRoomAttachments({ roomId, limit, offset, thumbnail }) {
      try {
        const response = await client.listRoomAttachments(
          {
            roomId,
            page: { limit, offset },
            thumbnail: thumbnailOptions(thumbnail)
          },
          { headers: headers() }
        );
        return {
          items: response.items.map(roomFileItem),
          totalCount: Number(response.page?.totalCount ?? 0),
          hasMore: response.page?.hasMore ?? false
        };
      } catch (err) {
        return handleAuthError(err);
      }
    },
    async refreshMessageAttachmentUrls(roomId, eventId, thumbnail) {
      try {
        const response = await client.refreshMessageAttachmentUrls(
          {
            roomId,
            eventId,
            thumbnail: thumbnailOptions(thumbnail)
          },
          { headers: headers() }
        );
        return new Map(
          response.attachments.map((attachment) => [
            attachment.attachmentId,
            {
              assetUrl: assetUrl(attachment.assetUrl) ?? { url: '', expiresAt: '' },
              thumbnailAssetUrl: assetUrl(attachment.thumbnailAssetUrl),
              videoThumbnailAssetUrl: assetUrl(attachment.videoThumbnailAssetUrl),
              variantAssetUrls: new Map(
                attachment.variants
                  .map((variant) => [variant.quality, assetUrl(variant.assetUrl)] as const)
                  .filter((entry): entry is readonly [string, ExpiringAssetUrl] => entry[1] !== null)
              )
            }
          ])
        );
      } catch (err) {
        return handleAuthError(err);
      }
    }
  };
}

function thumbnailOptions(options: AttachmentRefreshOptions): AttachmentThumbnailOptions {
  return new AttachmentThumbnailOptions({
    width: options.width,
    height: options.height,
    fit: options.fit === FitMode.Contain ? AttachmentFitMode.CONTAIN : AttachmentFitMode.COVER
  });
}

function roomFileItem(item: {
  messageEventId: string;
  threadRootEventId: string;
  createdAt?: { toDate(): Date };
  attachment?: RoomTimelineAttachment;
}): RoomFileItem {
  return {
    messageEventId: item.messageEventId,
    threadRootEventId: item.threadRootEventId || null,
    createdAt: timestampToISO(item.createdAt),
    attachment: attachment(item.attachment)
  };
}

function attachment(value?: RoomTimelineAttachment): RoomFileItem['attachment'] {
  return {
    id: value?.id ?? '',
    filename: value?.filename ?? '',
    contentType: value?.contentType ?? '',
    width: value?.width ?? 0,
    height: value?.height ?? 0,
    assetUrl: assetUrl(value?.assetUrl) ?? { url: '', expiresAt: '' },
    thumbnailAssetUrl: assetUrl(value?.thumbnailAssetUrl),
    videoProcessing: videoProcessing(value?.videoProcessing)
  };
}

function videoProcessing(
  value?: RoomTimelineVideoProcessing
): NonNullable<RoomFileItem['attachment']['videoProcessing']> | null {
  if (!value) return null;
  const status = videoProcessingStatus(value.status);
  if (!status) return null;
  return {
    status,
    durationMs: Number(value.durationMs) || null,
    width: value.width || null,
    height: value.height || null,
    sourceAvailable: value.sourceAvailable,
    reasonCode: value.reasonCode || null,
    thumbnailAssetUrl: assetUrl(value.thumbnailAssetUrl),
    variants: value.variants.map((variant) => ({
      quality: variant.quality,
      width: variant.width,
      height: variant.height,
      size: Number(variant.size),
      assetUrl: assetUrl(variant.assetUrl) ?? { url: '', expiresAt: '' }
    }))
  };
}

function videoProcessingStatus(
  status: RoomTimelineVideoProcessingStatus
): NonNullable<RoomFileItem['attachment']['videoProcessing']>['status'] | null {
  switch (status) {
    case RoomTimelineVideoProcessingStatus.PROCESSING:
      return 'PROCESSING';
    case RoomTimelineVideoProcessingStatus.COMPLETED:
      return 'COMPLETED';
    case RoomTimelineVideoProcessingStatus.FAILED:
      return 'FAILED';
    default:
      return null;
  }
}

function assetUrl(value?: RoomTimelineAssetUrl): ExpiringAssetUrl | null {
  if (!value) return null;
  return {
    url: value.url,
    expiresAt: timestampToISO(value.expiresAt)
  };
}

function timestampToISO(timestamp: { toDate(): Date } | undefined): string {
  return timestamp ? timestamp.toDate().toISOString() : '';
}
